### ISSUE TITLE
InvalidArgumentException: server version is invalid

### ISSUE THREAD

**Issue by anoymouserver** [nextcloud/server#25137]
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

### Steps to reproduce
1. run `sudo -u www-data ./occ app:update --allow-unstable news` (or `breezedark` or probably any other store app)
2. nothing happens but shortly after the log contains numerous `InvalidArgumentException: server version is invalid`
3.

### Expected behaviour
The app should update to the latest unstable/nightly.

### Actual behaviour
App is not updated but exeption is logged instead.

### Server configuration

**Operating system:** Debian 9.13

**Web server:** Nginx

**Database:** MariaDB 10.1.47

**PHP version:** 7.3.25

**Nextcloud version:** 21.0.0 beta6

**Updated from an older Nextcloud/ownCloud or fresh install:** updated from 21.0.0 beta5

**Where did you install Nextcloud from:** originally from the ZIP archive years ago

**Signing status:**
<details>
<summary>Signing status</summary>

```
No errors have been found.
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
Enabled:
  - accessibility: 1.7.0
  - activity: 2.14.3
  - admin_audit: 1.11.0
  - cloud_federation_api: 1.4.0
  - dav: 1.17.1
  - federatedfilesharing: 1.11.0
  - federation: 1.11.0
  - files: 1.16.0
  - files_markdown: 2.3.1
  - files_pdfviewer: 2.1.0
  - files_rightclick: 1.0.0
  - files_sharing: 1.13.1
  - files_texteditor: 2.14.0
  - files_trashbin: 1.11.0
  - files_versions: 1.14.0
  - files_videoplayer: 1.10.0
  - firstrunwizard: 2.10.0
  - logreader: 2.6.0
  - lookup_server_connector: 1.9.0
  - news: 15.1.1
  - notifications: 2.9.0
  - oauth2: 1.9.0
  - password_policy: 1.11.0
  - photos: 1.3.0
  - privacy: 1.5.0
  - provisioning_api: 1.11.0
  - serverinfo: 1.11.0
  - settings: 1.3.0
  - sharebymail: 1.11.0
  - theming: 1.12.0
  - twofactor_backupcodes: 1.10.0
  - updatenotification: 1.11.0
  - viewer: 1.5.0
  - workflowengine: 2.3.0
```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": ["cloud.example.com"],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "overwrite.cli.url": "https:\/\/cloud.example.com",
        "dbtype": "mysql",
        "version": "21.0.0.13",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "logtimezone": "UTC",
        "installed": true,
        "mail_smtpmode": "smtp",
        "theme": "",
        "loglevel": 2,
        "maintenance": false,
        "htaccess.RewriteBase": "\/",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "updater.release.channel": "beta",
        "mysql.utf8mb4": true,
        "memcache.local": "\\OC\\Memcache\\APCu",
        "knowledgebaseenabled": false,
        "token_auth_enforced": true,
        "trashbin_retention_obligation": "auto, 60",
        "app_install_overwrite": ["news", "files_texteditor", "files_markdown"],
        "mail_smtpsecure": "tls",
        "mail_sendmailmode": "smtp",
        "mail_smtpport": "587",
        "default_phone_region": "DE"
    }
}

```
</details>

**Are you using external storage, if yes which one:** no

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** no

### Client configuration
**Browser:** Not relevant

**Operating system:** Not relevant

#### Web server error log
<details>
<summary>Web server error log</summary>

```
nothing recent
```
</details>

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>

```
{
  "reqId": "TlnCKGnVyeBzk2edgAw6",
  "level": 2,
  "time": "2021-01-14T18:59:54+00:00",
  "remoteAddr": "",
  "user": "--",
  "app": "appstoreFetcher",
  "method": "",
  "url": "--",
  "message": {
    "Exception": "InvalidArgumentException",
    "Message": "server version is invalid",
    "Code": 0,
    "Trace": [
      {
        "file": "/var/www/cloud/lib/private/App/AppStore/Fetcher/AppFetcher.php",
        "line": 119,
        "function": "isCompatible",
        "class": "OC\\App\\CompareVersion",
        "type": "->",
        "args": [
          "7.3.25-1+0~20201130.73+debian9~1.gbp042074",
          "5.6",
          ">="
        ]
      },
      {
        "file": "/var/www/cloud/lib/private/App/AppStore/Fetcher/Fetcher.php",
        "line": 188,
        "function": "fetch",
        "class": "OC\\App\\AppStore\\Fetcher\\AppFetcher",
        "type": "->",
        "args": [
          "",
          "",
          true
        ]
      },
      {
        "file": "/var/www/cloud/lib/private/Installer.php",
        "line": 419,
        "function": "get",
        "class": "OC\\App\\AppStore\\Fetcher\\Fetcher",
        "type": "->",
        "args": [
          true
        ]
      },
      {
        "file": "/var/www/cloud/core/Command/App/Update.php",
        "line": 110,
        "function": "isUpdateAvailable",
        "class": "OC\\Installer",
        "type": "->",
        "args": [
          "breezedark",
          true
        ]
      },
      {
        "file": "/var/www/cloud/3rdparty/symfony/console/Command/Command.php",
        "line": 255,
        "function": "execute",
        "class": "OC\\Core\\Command\\App\\Update",
        "type": "->",
        "args": [
          {
            "__class__": "Symfony\\Component\\Console\\Input\\ArgvInput"
          },
          {
            "__class__": "Symfony\\Component\\Console\\Output\\ConsoleOutput"
          }
        ]
      },
      {
        "file": "/var/www/cloud/3rdparty/symfony/console/Application.php",
        "line": 1009,
        "function": "run",
        "class": "Symfony\\Component\\Console\\Command\\Command",
        "type": "->",
        "args": [
          {
            "__class__": "Symfony\\Component\\Console\\Input\\ArgvInput"
          },
          {
            "__class__": "Symfony\\Component\\Console\\Output\\ConsoleOutput"
          }
        ]
      },
      {
        "file": "/var/www/cloud/3rdparty/symfony/console/Application.php",
        "line": 273,
        "function": "doRunCommand",
        "class": "Symfony\\Component\\Console\\Application",
        "type": "->",
        "args": [
          {
            "__class__": "OC\\Core\\Command\\App\\Update"
          },
          {
            "__class__": "Symfony\\Component\\Console\\Input\\ArgvInput"
          },
          {
            "__class__": "Symfony\\Component\\Console\\Output\\ConsoleOutput"
          }
        ]
      },
      {
        "file": "/var/www/cloud/3rdparty/symfony/console/Application.php",
        "line": 149,
        "function": "doRun",
        "class": "Symfony\\Component\\Console\\Application",
        "type": "->",
        "args": [
          {
            "__class__": "Symfony\\Component\\Console\\Input\\ArgvInput"
          },
          {
            "__class__": "Symfony\\Component\\Console\\Output\\ConsoleOutput"
          }
        ]
      },
      {
        "file": "/var/www/cloud/lib/private/Console/Application.php",
        "line": 215,
        "function": "run",
        "class": "Symfony\\Component\\Console\\Application",
        "type": "->",
        "args": [
          {
            "__class__": "Symfony\\Component\\Console\\Input\\ArgvInput"
          },
          {
            "__class__": "Symfony\\Component\\Console\\Output\\ConsoleOutput"
          }
        ]
      },
      {
        "file": "/var/www/cloud/console.php",
        "line": 100,
        "function": "run",
        "class": "OC\\Console\\Application",
        "type": "->",
        "args": []
      },
      {
        "file": "/var/www/cloud/occ",
        "line": 11,
        "args": [
          "/var/www/cloud/console.php"
        ],
        "function": "require_once"
      }
    ],
    "File": "/var/www/cloud/lib/private/App/CompareVersion.php",
    "Line": 49,
    "CustomMessage": "--"
  },
  "userAgent": "--",
  "version": "21.0.0.13",
  "id": "600099d43761a"
}
{
  "reqId": "TlnCKGnVyeBzk2edgAw6",
  "level": 2,
  "time": "2021-01-14T18:59:54+00:00",
  "remoteAddr": "",
  "user": "--",
  "app": "appstoreFetcher",
  "method": "",
  "url": "--",
  "message": {
    "Exception": "InvalidArgumentException",
    "Message": "server version is invalid",
    "Code": 0,
    "Trace": [
      {
        "file": "/var/www/cloud/lib/private/App/AppStore/Fetcher/AppFetcher.php",
        "line": 119,
        "function": "isCompatible",
        "class": "OC\\App\\CompareVersion",
        "type": "->",
        "args": [
          "7.3.25-1+0~20201130.73+debian9~1.gbp042074",
          "5.6",
          ">="
        ]
      },
      {
        "file": "/var/www/cloud/lib/private/App/AppStore/Fetcher/Fetcher.php",
        "line": 188,
        "function": "fetch",
        "class": "OC\\App\\AppStore\\Fetcher\\AppFetcher",
        "type": "->",
        "args": [
          "",
          "",
          true
        ]
      },
      {
        "file": "/var/www/cloud/lib/private/Installer.php",
        "line": 419,
        "function": "get",
        "class": "OC\\App\\AppStore\\Fetcher\\Fetcher",
        "type": "->",
        "args": [
          true
        ]
      },
      {
        "file": "/var/www/cloud/core/Command/App/Update.php",
        "line": 110,
        "function": "isUpdateAvailable",
        "class": "OC\\Installer",
        "type": "->",
        "args": [
          "breezedark",
          true
        ]
      },
      {
        "file": "/var/www/cloud/3rdparty/symfony/console/Command/Command.php",
        "line": 255,
        "function": "execute",
        "class": "OC\\Core\\Command\\App\\Update",
        "type": "->",
        "args": [
          {
            "__class__": "Symfony\\Component\\Console\\Input\\ArgvInput"
          },
          {
            "__class__": "Symfony\\Component\\Console\\Output\\ConsoleOutput"
          }
        ]
      },
      {
        "file": "/var/www/cloud/3rdparty/symfony/console/Application.php",
        "line": 1009,
        "function": "run",
        "class": "Symfony\\Component\\Console\\Command\\Command",
        "type": "->",
        "args": [
          {
            "__class__": "Symfony\\Component\\Console\\Input\\ArgvInput"
          },
          {
            "__class__": "Symfony\\Component\\Console\\Output\\ConsoleOutput"
          }
        ]
      },
      {
        "file": "/var/www/cloud/3rdparty/symfony/console/Application.php",
        "line": 273,
        "function": "doRunCommand",
        "class": "Symfony\\Component\\Console\\Application",
        "type": "->",
        "args": [
          {
            "__class__": "OC\\Core\\Command\\App\\Update"
          },
          {
            "__class__": "Symfony\\Component\\Console\\Input\\ArgvInput"
          },
          {
            "__class__": "Symfony\\Component\\Console\\Output\\ConsoleOutput"
          }
        ]
      },
      {
        "file": "/var/www/cloud/3rdparty/symfony/console/Application.php",
        "line": 149,
        "function": "doRun",
        "class": "Symfony\\Component\\Console\\Application",
        "type": "->",
        "args": [
          {
            "__class__": "Symfony\\Component\\Console\\Input\\ArgvInput"
          },
          {
            "__class__": "Symfony\\Component\\Console\\Output\\ConsoleOutput"
          }
        ]
      },
      {
        "file": "/var/www/cloud/lib/private/Console/Application.php",
        "line": 215,
        "function": "run",
        "class": "Symfony\\Component\\Console\\Application",
        "type": "->",
        "args": [
          {
            "__class__": "Symfony\\Component\\Console\\Input\\ArgvInput"
          },
          {
            "__class__": "Symfony\\Component\\Console\\Output\\ConsoleOutput"
          }
        ]
      },
      {
        "file": "/var/www/cloud/console.php",
        "line": 100,
        "function": "run",
        "class": "OC\\Console\\Application",
        "type": "->",
        "args": []
      },
      {
        "file": "/var/www/cloud/occ",
        "line": 11,
        "args": [
          "/var/www/cloud/console.php"
        ],
        "function": "require_once"
      }
    ],
    "File": "/var/www/cloud/lib/private/App/CompareVersion.php",
    "Line": 49,
    "CustomMessage": "--"
  },
  "userAgent": "--",
  "version": "21.0.0.13",
  "id": "600099d4376e5"
}
{
  "reqId": "TlnCKGnVyeBzk2edgAw6",
  "level": 2,
  "time": "2021-01-14T18:59:54+00:00",
  "remoteAddr": "",
  "user": "--",
  "app": "appstoreFetcher",
  "method": "",
  "url": "--",
  "message": {
    "Exception": "InvalidArgumentException",
    "Message": "server version is invalid",
    "Code": 0,
    "Trace": [
      {
        "file": "/var/www/cloud/lib/private/App/AppStore/Fetcher/AppFetcher.php",
        "line": 119,
        "function": "isCompatible",
        "class": "OC\\App\\CompareVersion",
        "type": "->",
        "args": [
          "7.3.25-1+0~20201130.73+debian9~1.gbp042074",
          "5.6",
          ">="
        ]
      },
      {
        "file": "/var/www/cloud/lib/private/App/AppStore/Fetcher/Fetcher.php",
        "line": 188,
        "function": "fetch",
        "class": "OC\\App\\AppStore\\Fetcher\\AppFetcher",
        "type": "->",
        "args": [
          "",
          "",
          true
        ]
      },
      {
        "file": "/var/www/cloud/lib/private/Installer.php",
        "line": 419,
        "function": "get",
        "class": "OC\\App\\AppStore\\Fetcher\\Fetcher",
        "type": "->",
        "args": [
          true
        ]
      },
      {
        "file": "/var/www/cloud/core/Command/App/Update.php",
        "line": 110,
        "function": "isUpdateAvailable",
        "class": "OC\\Installer",
        "type": "->",
        "args": [
          "breezedark",
          true
        ]
      },
      {
        "file": "/var/www/cloud/3rdparty/symfony/console/Command/Command.php",
        "line": 255,
        "function": "execute",
        "class": "OC\\Core\\Command\\App\\Update",
        "type": "->",
        "args": [
          {
            "__class__": "Symfony\\Component\\Console\\Input\\ArgvInput"
          },
          {
            "__class__": "Symfony\\Component\\Console\\Output\\ConsoleOutput"
          }
        ]
      },
      {
        "file": "/var/www/cloud/3rdparty/symfony/console/Application.php",
        "line": 1009,
        "function": "run",
        "class": "Symfony\\Component\\Console\\Command\\Command",
        "type": "->",
        "args": [
          {
            "__class__": "Symfony\\Component\\Console\\Input\\ArgvInput"
          },
          {
            "__class__": "Symfony\\Component\\Console\\Output\\ConsoleOutput"
          }
        ]
      },
      {
        "file": "/var/www/cloud/3rdparty/symfony/console/Application.php",
        "line": 273,
        "function": "doRunCommand",
        "class": "Symfony\\Component\\Console\\Application",
        "type": "->",
        "args": [
          {
            "__class__": "OC\\Core\\Command\\App\\Update"
          },
          {
            "__class__": "Symfony\\Component\\Console\\Input\\ArgvInput"
          },
          {
            "__class__": "Symfony\\Component\\Console\\Output\\ConsoleOutput"
          }
        ]
      },
      {
        "file": "/var/www/cloud/3rdparty/symfony/console/Application.php",
        "line": 149,
        "function": "doRun",
        "class": "Symfony\\Component\\Console\\Application",
        "type": "->",
        "args": [
          {
            "__class__": "Symfony\\Component\\Console\\Input\\ArgvInput"
          },
          {
            "__class__": "Symfony\\Component\\Console\\Output\\ConsoleOutput"
          }
        ]
      },
      {
        "file": "/var/www/cloud/lib/private/Console/Application.php",
        "line": 215,
        "function": "run",
        "class": "Symfony\\Component\\Console\\Application",
        "type": "->",
        "args": [
          {
            "__class__": "Symfony\\Component\\Console\\Input\\ArgvInput"
          },
          {
            "__class__": "Symfony\\Component\\Console\\Output\\ConsoleOutput"
          }
        ]
      },
      {
        "file": "/var/www/cloud/console.php",
        "line": 100,
        "function": "run",
        "class": "OC\\Console\\Application",
        "type": "->",
        "args": []
      },
      {
        "file": "/var/www/cloud/occ",
        "line": 11,
        "args": [
          "/var/www/cloud/console.php"
        ],
        "function": "require_once"
      }
    ],
    "File": "/var/www/cloud/lib/private/App/CompareVersion.php",
    "Line": 49,
    "CustomMessage": "--"
  },
  "userAgent": "--",
  "version": "21.0.0.13",
  "id": "600099d4377b0"
}
{
  "reqId": "TlnCKGnVyeBzk2edgAw6",
  "level": 2,
  "time": "2021-01-14T18:59:54+00:00",
  "remoteAddr": "",
  "user": "--",
  "app": "appstoreFetcher",
  "method": "",
  "url": "--",
  "message": {
    "Exception": "InvalidArgumentException",
    "Message": "server version is invalid",
    "Code": 0,
    "Trace": [
      {
        "file": "/var/www/cloud/lib/private/App/AppStore/Fetcher/AppFetcher.php",
        "line": 119,
        "function": "isCompatible",
        "class": "OC\\App\\CompareVersion",
        "type": "->",
        "args": [
          "7.3.25-1+0~20201130.73+debian9~1.gbp042074",
          "5.6",
          ">="
        ]
      },
      {
        "file": "/var/www/cloud/lib/private/App/AppStore/Fetcher/Fetcher.php",
        "line": 188,
        "function": "fetch",
        "class": "OC\\App\\AppStore\\Fetcher\\AppFetcher",
        "type": "->",
        "args": [
          "",
          "",
          true
        ]
      },
      {
        "file": "/var/www/cloud/lib/private/Installer.php",
        "line": 419,
        "function": "get",
        "class": "OC\\App\\AppStore\\Fetcher\\Fetcher",
        "type": "->",
        "args": [
          true
        ]
      },
      {
        "file": "/var/www/cloud/core/Command/App/Update.php",
        "line": 110,
        "function": "isUpdateAvailable",
        "class": "OC\\Installer",
        "type": "->",
        "args": [
          "breezedark",
          true
        ]
      },
      {
        "file": "/var/www/cloud/3rdparty/symfony/console/Command/Command.php",
        "line": 255,
        "function": "execute",
        "class": "OC\\Core\\Command\\App\\Update",
        "type": "->",
        "args": [
          {
            "__class__": "Symfony\\Component\\Console\\Input\\ArgvInput"
          },
          {
            "__class__": "Symfony\\Component\\Console\\Output\\ConsoleOutput"
          }
        ]
      },
      {
        "file": "/var/www/cloud/3rdparty/symfony/console/Application.php",
        "line": 1009,
        "function": "run",
        "class": "Symfony\\Component\\Console\\Command\\Command",
        "type": "->",
        "args": [
          {
            "__class__": "Symfony\\Component\\Console\\Input\\ArgvInput"
          },
          {
            "__class__": "Symfony\\Component\\Console\\Output\\ConsoleOutput"
          }
        ]
      },
      {
        "file": "/var/www/cloud/3rdparty/symfony/console/Application.php",
        "line": 273,
        "function": "doRunCommand",
        "class": "Symfony\\Component\\Console\\Application",
        "type": "->",
        "args": [
          {
            "__class__": "OC\\Core\\Command\\App\\Update"
          },
          {
            "__class__": "Symfony\\Component\\Console\\Input\\ArgvInput"
          },
          {
            "__class__": "Symfony\\Component\\Console\\Output\\ConsoleOutput"
          }
        ]
      },
      {
        "file": "/var/www/cloud/3rdparty/symfony/console/Application.php",
        "line": 149,
        "function": "doRun",
        "class": "Symfony\\Component\\Console\\Application",
        "type": "->",
        "args": [
          {
            "__class__": "Symfony\\Component\\Console\\Input\\ArgvInput"
          },
          {
            "__class__": "Symfony\\Component\\Console\\Output\\ConsoleOutput"
          }
        ]
      },
      {
        "file": "/var/www/cloud/lib/private/Console/Application.php",
        "line": 215,
        "function": "run",
        "class": "Symfony\\Component\\Console\\Application",
        "type": "->",
        "args": [
          {
            "__class__": "Symfony\\Component\\Console\\Input\\ArgvInput"
          },
          {
            "__class__": "Symfony\\Component\\Console\\Output\\ConsoleOutput"
          }
        ]
      },
      {
        "file": "/var/www/cloud/console.php",
        "line": 100,
        "function": "run",
        "class": "OC\\Console\\Application",
        "type": "->",
        "args": []
      },
      {
        "file": "/var/www/cloud/occ",
        "line": 11,
        "args": [
          "/var/www/cloud/console.php"
        ],
        "function": "require_once"
      }
    ],
    "File": "/var/www/cloud/lib/private/App/CompareVersion.php",
    "Line": 49,
    "CustomMessage": "--"
  },
  "userAgent": "--",
  "version": "21.0.0.13",
  "id": "600099d43787a"
}

... and the same on and on and on over pages
```
</details>

#### Browser log
<details>
<summary>Browser log</summary>

```
not relevant -> happens on commandline
```
</details>

**Comment by gouttegd** [comment]
I observe the same problem with Nextcloud 20.0.5.

I believe it is caused by commit 24237f1a34b17dbf819403179577d6cc178de313 (backported in the 20.0.x branch with commit 0ae5c8dd3b88e1eb22981815cc1e86a478047772), which reuses the `OC\App\CompareVersion::isCompatible` method to compare PHP versions even though that method seems intended to compare Nextcloud versions only.

**Comment by kesselb** [comment]
> I believe it is caused by commit [24237f1](https://github.com/nextcloud/server/commit/24237f1a34b17dbf819403179577d6cc178de313) (backported in the 20.0.x branch with commit [0ae5c8d](https://github.com/nextcloud/server/commit/0ae5c8dd3b88e1eb22981815cc1e86a478047772)), which reuses the `OC\App\CompareVersion::isCompatible` method to compare PHP versions even though that method seems intended to compare Nextcloud versions only.

I would agree. Mind to sent a pull request? I guess https://www.php.net/manual/en/function.version-compare.php should work here.

**Comment by gouttegd** [comment]
`version_compare` seems to do the job indeed, a pull request is on the way.

**Comment by kesselb** [comment]
Fixed by #25138
